json: Compile crash when json.encode

### Describe the bug

C compile crash saying: error: struct or union expected on the code below on line 7877. I see that it is trying to access a property via the "." instead of "->" while it should use "->". This code was working fine yesterday and fails since I upgraded to latest v version.

This is the Node struct:
```
pub enum PlatformType {
	unknown
	osx
	ubuntu
	alpine
}

pub enum CPUType {
	unknown
	intel
	arm
	intel32
	arm32
}

[heap]
pub struct Node {
mut:
	executor Executor        [skip; str: skip]
	factory  &BuilderFactory [skip; str: skip]
pub:
	name string = 'mymachine'
pub mut:
	platform    PlatformType
	cputype     CPUType
	done        map[string]string
	environment map[string]string
	params      Params
}
```


![image](https://user-images.githubusercontent.com/7792337/225887327-267434a4-1b41-492a-b48f-59dca7056f61.png)


### Expected Behavior

See description

### Current Behavior

See description

### Reproduction Steps

See description

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.3.3 6e1e406

### Environment details (OS name and version, etc.)

V full version: V 0.3.3 5454562.6e1e406
OS: linux, Ubuntu 22.04.1 LTS
Processor: 16 cpus, 64bit, little endian, 11th Gen Intel(R) Core(TM) i7-11800H @ 2.30GHz

getwd: /work/crystallib
vexe: /work/v/v
vexe mtime: 2023-03-17 10:44:32

vroot: OK, value: /work/v
VMODULES: OK, value: /home/brandon/.vmodules
VTMP: OK, value: /tmp/v_1000

Git version: git version 2.34.1
Git vroot status: weekly.2023.10-70-g6e1e4062
.git/config present: true

CC version: cc (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
thirdparty/tcc status: thirdparty-linux-amd64 12f392c3

With `V 0.3.3 6e1e406` I get the following:
```
$ v -g foo.v
foo.v:32:2: warning: unused variable: `data`
   30 | 
   31 | pub fn (mut node Node) save() ! {
   32 |     data :=  json.encode(node)
      |     ~~~~
   33 | }
   34 |
/tmp/v_1000/foo.14519348190291310068.tmp.c:2286: error: struct or union expected
builder error: 
==================
C error. This should never happen.
```

--- a/vlib/json/json_encode_with_mut_test.v
+++ b/vlib/json/json_encode_with_mut_test.v
@@ -1,0 +1,42 @@
+module main
+
+import json
+
+pub enum PlatformType {
+	unknown
+	osx
+	ubuntu
+	alpine
+}
+
+pub enum CPUType {
+	unknown
+	intel
+	arm
+	intel32
+	arm32
+}
+
+[heap]
+pub struct Node {
+pub:
+	name string = 'mymachine'
+pub mut:
+	platform    PlatformType
+	cputype     CPUType
+	done        map[string]string
+	environment map[string]string
+}
+
+pub fn (mut node Node) save() ! {
+	data := json.encode(node)
+	dump(data)
+}
+
+fn test_encode_with_mut_struct() {
+	mut n := Node{
+		platform: .osx
+		cputype: .unknown
+	}
+	n.save() or { panic(err) }
+}

--- a/vlib/v/gen/c/json.v
+++ b/vlib/v/gen/c/json.v
@@ -520,21 +520,25 @@ fn (mut g Gen) gen_struct_enc_dec(utyp ast.Type, type_info ast.TypeInfo, styp st
 			}
 		}
 		if field_sym.kind == .enum_ {
-			enc.writeln('\tcJSON_AddItemToObject(o, "${name}", json__encode_u64(${prefix_enc}.${c_name(field.name)}));\n')
+			enc.writeln('\tcJSON_AddItemToObject(o, "${name}", json__encode_u64(${prefix_enc}${op}${c_name(field.name)}));\n')
 		} else {
 			if field_sym.name == 'time.Time' {
 				// time struct requires special treatment
 				// it has to be encoded as a unix timestamp number
-				enc.writeln('\tcJSON_AddItemToObject(o, "${name}", json__encode_u64(${prefix_enc}.${c_name(field.name)}._v_unix));')
+				enc.writeln('\tcJSON_AddItemToObject(o, "${name}", json__encode_u64(${prefix_enc}${op}${c_name(field.name)}._v_unix));')
 			} else {
 				if !field.typ.is_real_pointer() {
 					enc.writeln('\tcJSON_AddItemToObject(o, "${name}", ${enc_name}(${prefix_enc}${op}${c_name(field.name)})); /*A*/')
 				} else {
 					arg_prefix := if field.typ.is_ptr() { '' } else { '*' }
 					sptr_value := '${prefix_enc}${op}${c_name(field.name)}'
-					enc.writeln('\tif (${sptr_value} != 0) {')
-					enc.writeln('\t\tcJSON_AddItemToObject(o, "${name}", ${enc_name}(${arg_prefix}${sptr_value}));')
-					enc.writeln('\t}\n')
+					if !field.typ.has_flag(.option) {
+						enc.writeln('\tif (${sptr_value} != 0) {')
+						enc.writeln('\t\tcJSON_AddItemToObject(o, "${name}", ${enc_name}(${arg_prefix}${sptr_value}));')
+						enc.writeln('\t}\n')
+					} else {
+						enc.writeln('\t\tcJSON_AddItemToObject(o, "${name}", ${enc_name}(${arg_prefix}${sptr_value}));')
+					}
 				}
 			}
 		}

